hist/histv7/doc/CodeArchitecture.md

```diff
@@ -0,0 +1,75 @@
+# Code Architecture
+
+This document lists all classes, describes their responsibilities, and details how they fit together.
+
+## Core Classes
+
+These classes are involved when the user calls `Fill`.
+The list is ordered "bottom-up" in terms of functionality.
+
+### `RRegularAxis`, `RVariableBinAxis`, `RCategoricalAxis`
+
+These classes implement a concrete axis type.
+Their main task is to compute the linearized index for a single `Fill` argument:
+```c++
+std::pair<std::size_t, bool> ComputeLinearizedIndex(double x);
+```
+The `bool` is used to indicate if the return value is valid.
+For example, the argument may be outside the axis with the underflow and overflow bins disabled.
+`std::pair` is chosen over `std::optional` because it allows to return the same linearized index, even if invalid.
+
+### `Internal::RAxes`
+
+This class is responsible for managing a histogram's axis configuration.
+It stores the axis objects as a `std::vector` of `std::variant`s.
+Objects of this class are used internally and not exposed to the user.
+Relevant functionality is available through user-facing classes such as `RHistEngine`.
```

Comment:
Here I'm still unsure if we want this class in `Internal`, or in `Detail` and conditionally exposed to users. I think we don't have to decide immediately, but should keep in mind that (one day) the class name will be part of on-disk data...